commons-jetty/src/main/scala/com/avsystem/commons/jetty/rpc/JValueInputOutput.scala

```diff
@@ -0,0 +1,115 @@
+package jetty.rpc
+
+import java.util.Base64
+
+import com.avsystem.commons.serialization.GenCodec.{ReadFailure, WriteFailure}
+import com.avsystem.commons.serialization.{FieldInput, InputType, ListInput, ObjectInput, _}
+import jawn.ast.{JArray, JBool, JFalse, JNull, JNum, JObject, JString, JTrue, JValue}
+
+object JValueOutput {
+  def write[T: GenCodec](value: T): JValue = {
+    var result: Opt[JValue] = Opt.Empty
+    val output = new JValueOutput(value => result = Opt(value))
+    GenCodec.write(output, value)
+    result.getOrElse(throw new WriteFailure("No value written by GenCodec"))
+  }
+}
+
+final class JValueOutput(private val consumer: JValue => Unit) extends AnyVal with Output {
+  def writeNull() = consumer(JNull)
+  def writeBoolean(boolean: Boolean) = consumer(if (boolean) JTrue else JFalse)
+  def writeString(str: String) = consumer(JString(str))
+  def writeInt(int: Int) = consumer(JNum(int))
+  def writeLong(long: Long) = consumer(JNum(long))
+  def writeDouble(double: Double) = consumer(JNum(double))
+  def writeBinary(binary: Array[Byte]) = consumer(JString(Base64.getEncoder.encodeToString(binary)))
+  def writeList() = new JValueListOutput(consumer)
+  def writeObject() = new JValueObjectOutput(consumer)
+}
+
+final class JValueListOutput(consumer: JArray => Unit) extends ListOutput {
+  private val builder = Array.newBuilder[JValue]
+
+  def writeElement() = new JValueOutput(builder += _)
+  def finish() = consumer(JArray(builder.result()))
+}
+
+final class JValueObjectOutput(consumer: JObject => Unit) extends ObjectOutput {
+  private val builder = MMap.newBuilder[String, JValue]
+
+  def writeField(key: String) = new JValueOutput(v => builder += ((key, v)))
+  def finish() = consumer(JObject(builder.result()))
+}
+
+object JValueInput {
+  def read[T: GenCodec](value: JValue): T =
+    GenCodec.read[T](new JValueInput(value))
+}
+
+class JValueInput(value: JValue) extends Input {
+
+  def inputType: InputType = value match {
+    case _: JArray => InputType.List
+    case _: JObject => InputType.Object
+    case JNull => InputType.Null
+    case _ => InputType.Simple
+  }
+
+  private def expectedError(expected: String) =
+    throw new ReadFailure(s"Expected JSON $expected but got ${value.valueType}")
+
+  def readNull() = value match {
+    case JNull => null
+    case _ => expectedError("null")
+  }
+  def readBoolean() = value match {
+    case v: JBool => v.asBoolean
+    case _ => expectedError("boolean")
+  }
+  def readString() = value match {
+    case JString(s) => s
+    case _ => expectedError("string")
+  }
+  def readBinary() = value match {
+    case JString(s) => try Base64.getDecoder.decode(s) catch {
+      case e: IllegalArgumentException => throw new ReadFailure("error decoding Base64: " + e.getMessage, e)
+    }
+    case _ => expectedError("string")
+  }
+  def readInt() = value match {
+    case num: JNum => Try(num.asInt).getOrElse(throw new ReadFailure(s"not an int: $num"))
```

Comment:
`asInt` and `asLong` calls can fail only for `DeferNum`/`DeferLong`, although it's highly unlikely, as the Jawn parser shouldn't produce such values. The same goes for `asDouble`. No need to treat those differently - either all three should use `Try`, or neither should.